File: remote/src/main/java/com/kelsos/mbrc/presenters/BrowseAlbumPresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.MetaDataType;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.domain.Album;
import com.kelsos.mbrc.interactors.LibraryAlbumInteractor;
import com.kelsos.mbrc.interactors.QueueInteractor;
import com.kelsos.mbrc.ui.views.BrowseAlbumView;
import roboguice.util.Ln;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static com.kelsos.mbrc.constants.Constants.PAGE_SIZE;

public class BrowseAlbumPresenterImpl implements BrowseAlbumPresenter {

  @Inject private QueueInteractor queueInteractor;
  @Inject private LibraryAlbumInteractor albumInteractor;
  private BrowseAlbumView view;

  @Override public void bind(BrowseAlbumView view) {
    this.view = view;
  }

  @Override public void queue(Album album, @Queue.Action String action) {
    queueInteractor.execute(MetaDataType.ALBUM, action, (int) album.getId()).subscribe(aBoolean -> {

    }, Ln::v);
  }

  @Override public void load() {
    albumInteractor.execute(0, PAGE_SIZE)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(albums -> {
          view.clearData();
          view.updateData(albums);
        }, Ln::v);
  }

  @Override public void load(int page) {
    albumInteractor.execute(page * PAGE_SIZE, PAGE_SIZE)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(albums -> view.updateData(albums), Ln::v);
  }
}

File: remote/src/main/kotlin/com/kelsos/mbrc/presenters/BrowseAlbumPresenterImpl.kt
package com.kelsos.mbrc.presenters

import com.google.inject.Inject
import com.kelsos.mbrc.annotations.MetaDataType
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.constants.Constants.PAGE_SIZE
import com.kelsos.mbrc.domain.Album
import com.kelsos.mbrc.interactors.LibraryAlbumInteractor
import com.kelsos.mbrc.interactors.QueueInteractor
import com.kelsos.mbrc.ui.views.BrowseAlbumView
import roboguice.util.Ln
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class BrowseAlbumPresenterImpl : BrowseAlbumPresenter {

  @Inject private lateinit var queueInteractor: QueueInteractor
  @Inject private lateinit var albumInteractor: LibraryAlbumInteractor
  private var view: BrowseAlbumView? = null

  override fun bind(view: BrowseAlbumView) {
    this.view = view
  }

  override fun queue(album: Album, @Queue.Action action: String) {
    queueInteractor.execute(MetaDataType.ALBUM, action, album.id.toInt()).subscribe({

    }, { Ln.v(it) })
  }

  override fun load() {
    albumInteractor.execute(0,
        PAGE_SIZE).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({
      view?.clearData()
      view?.updateData(it)
    }, { Ln.v(it) })
  }

  override fun load(page: Int) {
    albumInteractor.execute(page * PAGE_SIZE,
        PAGE_SIZE).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({
      view?.updateData(it)
    }, { Ln.v(it) })
  }
}
